### PR TITLE
 Make @preconcurrency import hide sendable variance

### PR DESCRIPTION
**Pull request by beccadax**
When the conformance checker and override checker detect a difference in a function type’s `@Sendable` attribute that varies in an illegal way, they now check if the protocol/base class was imported with an `@preconcurrency import`, and either limit the diagnostic or suggest adding `@preconcurrency` to the import as appropriate.

Additionally, this PR makes it so we don't emit the remark recommending that `@preconcurrency` be added to an import if the type we're using has explicit sendability; after all, the author of the type has audited it and determined that any sendability problems in your code are not an error. Types in `isConcurrencyChecked()` modules are treated as always having explicit sendability for these purposes, since eventually we want to eliminate `isConcurrencyChecked()` and just serialize their public types as though they are explicitly non-sendable.

Completes rdar://91109455. Followup for #42194.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 1]
```diff
@@ -757,82 +757,107 @@ DiagnosticBehavior SendableCheckContext::diagnosticBehavior(
...
-          nominal->getName());
+  bool wasSuppressed = diagnose(behavior);
+
+  bool emittedDiagnostics =
+      behavior != DiagnosticBehavior::Ignore && !wasSuppressed;
+
+  // When the type is explicitly Sendable *or* explicitly non-Sendable, we
+  // assume it has been audited and `@preconcurrency` is not recommended even
+  // though it would actually affect the diagnostic.
+  bool nominalIsImportedAndHasExplicitSendability =
```
Comment (beccadax) [inline lib/Sema/TypeCheckConcurrency.cpp:779]: Ack, this variable name should end with `ImplicitSendability`, not `Explicit`.

File: lib/Sema/TypeCheckConcurrency.h [hunk 2]
```diff
@@ -389,6 +391,26 @@ bool diagnoseNonSendableTypes(
...
+///
+/// \param fromContext The context where the error will be emitted.
+///
+/// \param diagnose Emit a diagnostic indicating a sendability problem,
+/// with the suggested behavior limitation. Returns \c true
+/// if it did not emit any diagnostics.
+///
+/// \returns \c true if any errors were produced, \c false if no diagnostics or
+/// only warnings and notes were produced.
+bool diagnoseAsNonSendableBasedOn(
```
Comment (beccadax) [inline lib/Sema/TypeCheckConcurrency.h:410]: This could probably use a better name—suggestions welcome.

Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.h:410]: `diagnoseAsNonSendableFromContext`? It's not that much better.

**Review by DougGregor — APPROVED** [review]
This looks great!

File: lib/Sema/TypeCheckConcurrency.h (hunk 2 above)
Comment (beccadax) [reply lib/Sema/TypeCheckConcurrency.h]: I ended up with `diagnoseSendabilityErrorBasedOn()`, which is at least more vague about the kind of error we're emitting.

**Comment by beccadax** [comment]
@swift-ci smoke test and merge

**Comment by beccadax** [comment]
@swift-ci please smoke test linux platform

**Comment by beccadax** [comment]
@swift-ci smoke test macOS platform

**Pull request merged**